Modify rule S5332: Improve description (#474)
SonarSource/rspec: rules/S5332/description.adoc

> Clear-text protocols such as ``++ftp++``, ``++telnet++`` or  non secure ``++http++`` are lacking encryption of transported data. They are also missing the capability to build an authenticated connection. This mean that any attacker who can sniff traffic from the network can read, modify or corrupt the transported content. These protocol are not secure as they expose applications to a large range of risk:

Reviewer: Do you think it could be interesting to maintain a list of unsafe protocols in the SonarQube docs that we could link here ? Because there are some "new"(||newly trending) protocols that can appear, and no be "Secure by default" (and I think we also should make a docs page about the AppSec principles and link it here for more informations). 

After that, we could maintain the protocol list just as we maintain rules

> Clear-text protocols such as ``++ftp++``, ``++telnet++`` or  non secure ``++http++`` are lacking encryption of transported data. They are also missing the capability to build an authenticated connection. This mean that any attacker who can sniff traffic from the network can read, modify or corrupt the transported content. These protocol are not secure as they expose applications to a large range of risk:

Reviewer: Do you think it could be interesting to maintain a list of unsafe protocols in the SonarQube docs that we could link here ? Because there are some "new"(||newly trending) protocols that can appear, and no be "Secure by default" (and I think we also should make a docs page about the AppSec principles and link it here for more informations). 

After that, we could maintain the protocol list just as we maintain rules

* Sensitive data exposure
* Traffic redirected  to a malicious endpoint
* Malware infected software update or installer
* Execution of client side code
* Corruption of critical information

> Cloud components that run in a logically isolated network should avoid using clear-text protocols whenever it's possible. The main reason for enabling encryption in-transit in such a case is to apply the principle of _defense in depth_. An attacker who would successfully bypass isolation mechanisms or take control of one of the components of the network, would be unable to sniff or temper with the transported if encryption in-transit is enabled.

Reviewer: I have diverse thoughts about this paragraph, and I made a bullet list but I'll just post an example of what I think, more understandable than my previous bullet points haha :

``` 
Note that even systems running in segmented networks should communicate with
encrypted protocols (Even in Internet-less environments or Cloud environments).
Layering multiple layers of basic security will help keep the system secure 
by thwarting attackers who were already able to compromise a defense
mechanism.

For example, an attacker who either successfully bypassed isolation mechanisms
or compromised a system may already by able to sniff or tamper with the
transported data. They may already be allowed to exploit a data breach, or even
to pivot in the Information System.

Securing systems by layering security basics is also known as applying the **Defense in Depth**
principle. (<-- link to our docs that could be more specific about this principle, and where we 
could add even more text, business impact, examples, just to be even more concise here)
```

Clear-text protocols such as ``++ftp++``, ``++telnet++`` or  non secure ``++http++`` are lacking encryption of transported data. They are also missing the capability to build an authenticated connection. This mean that any attacker who can sniff traffic from the network can read, modify or corrupt the transported content. These protocol are not secure as they expose applications to a large range of risk:

* Sensitive data exposure
* Traffic redirected  to a malicious endpoint
* Malware infected software update or installer
* Execution of client side code
* Corruption of critical information

> Even in a context of isolated networks like offline environments or cloud environments, there is a risk of communication being sniffed or tampered with. For example, if an attacker could successfully bypass isolation mechanisms or compromise a component of the network. In such a case, the attacker would use clear-text communications as a lever to exploit a data breach or to compromise other components of the network.
> This principle of securing systems by layering security basics is also known as _defense-in-depth_.

Reviewer: I am a noob and closed TWO times the tab while doing the review, so here is a sum-up of what I wanted to do : 
- Slowly introduce the defense in depth principle
  - I did it by introducing "layers" and emphasizing it like in ``compromise prior security layers`` or ``encrypting communications as another hardening layer``
- Add the "IAM compromise" scenario that could be relevant today (actual person, service account, etc.)
- be more concise (I may have failed on this one) 
- I thought that we went too quickly into communications being sniffed without saying "by whom", so I added a "why" example -> the insider threat. : So Why : "Insider threat", How : "Defense In Depth", What : Follow this rule

```suggestion
Even in a context of isolated networks like offline environments or segmented cloud environments, the insider threat exists. Thus, attacks involving communications being sniffed or tampered with still exist. 
For example, an attacker could successfully compromise prior security layers by :
- Bypassing isolation mechanisms,
- Compromising a component of the network,
- Getting the credentials of an internal IAM account (either from a service account or an actual person).

In such cases, encrypting communications as another hardening layer would prevent the attacker from exploiting clear-text communications. They won't be able to publicly leak data or to compromise other components of the network.
By layering diverse security practices (segmentation and encryption for example), the code will follow the _defense-in-depth_ principle.
```

> Clear-text protocols such as ``++ftp++``, ``++telnet++`` or  non-secure ``++http++`` lack encryption of transported data, as well as the capability to build an authenticated connection. 
> This means that any attacker able to sniff traffic from the network can read, modify or corrupt the transported content. These protocols are not secure as they expose applications to a large range of risk:

Reviewer: ```suggestion
Clear-text protocols such as ``++ftp++``, ``++telnet++`` or  non-secure ``++http++`` lack encryption of transported data, as well as the capability to build an authenticated connection. 
It means that an attacker able to sniff traffic from the network can read, modify or corrupt the transported content. These protocols are not secure as they expose applications to an extensive range of risks:
```

* Sensitive data exposure
* Traffic redirected  to a malicious endpoint
* Malware infected software update or installer
* Execution of client side code
* Corruption of critical information

Even in the context of isolated networks like offline environments or segmented cloud environments, the insider threat exists. Thus, attacks involving communications being sniffed or tampered with can still happen.

> For example, attackers could successfully compromise prior security layers by :

Reviewer 2: ```suggestion
For example, attackers could successfully compromise prior security layers by:
```

The French space :smile:

* Bypassing isolation mechanisms
* Compromising a component of the network
* Getting the credentials of an internal IAM account (either from a service account or an actual person)

> In such cases, encrypting communications would prevent attackers from exploiting clear-text communications. Indeed, this additional protection layer would prevent them from leaking data or compromising other network components.

Reviewer 2: I think the following part is not very precise at the moment:
> exploiting clear-text communications

Encrypting the communication does not prevent every form of exploitation (e.g., you can still exploit a SQL injection through HTTPS).

The `indeed` seems a bit out of place here.

Lastly, I am not sure how much encrypting transit prevents attackers from compromising other network components.

Author: @hendrik-buchwald-sonarsource  what about this
```suggestion
In such cases, encrypting communications would decrease the chances of attackers to successfully leak data or steal credentials from other network components.
```

> Even in the context of isolated networks like offline environments or segmented cloud environments, the insider threat exists. Thus, attacks involving communications being sniffed or tampered can still happen.
> 
> For example, attackers could successfully compromise prior security layers by :
> * Bypassing isolation mechanisms
> * Compromising a component of the network
> * Getting the credentials of an internal IAM account (either from a service account or an actual person)
> 
> In such cases, encrypting communications would prevent attackers from exploiting clear-text communications. This additional protection layer would prevent them from leaking data or compromising other components of the network.
> By layering various security practices (segmentation and encryption for example), the application will follow the _defense-in-depth_ principle.
> 
> Note also that using the ``++http++`` protocol is being deprecated by https://blog.mozilla.org/security/2015/04/30/deprecating-non-secure-http[major web browser].

Reviewer: ```suggestion
Even in the context of isolated networks like offline environments or segmented cloud environments, the insider threat exists. Thus, attacks involving communications being sniffed or tampered with can still happen.

For example, attackers could successfully compromise prior security layers by :
* Bypassing isolation mechanisms
* Compromising a component of the network
* Getting the credentials of an internal IAM account (either from a service account or an actual person)

In such cases, encrypting communications would prevent attackers from exploiting clear-text communications. Indeed, this additional protection layer would prevent them from leaking data or compromising other network components.
By layering various security practices (segmentation and encryption, for example), the application will follow the _defense-in-depth_ principle.

Note that using the ``++http++`` protocol is being deprecated by https://blog.mozilla.org/security/2015/04/30/deprecating-non-secure-http[major web browser].
```

> Note that using the ``++http++`` protocol is being deprecated by https://blog.mozilla.org/security/2015/04/30/deprecating-non-secure-http[major web browser].

Reviewer 2: ```suggestion
Note that using the ``++http++`` protocol is being deprecated by https://blog.mozilla.org/security/2015/04/30/deprecating-non-secure-http[major web browsers].
```
